charmhub_lp_tools/charm_project.py

```diff
@@ -812,6 +819,108 @@ def _do_authorization(self, recipe: TypeLPObject) -> None:
                 "Failed authenticating for upload.  Recipe: %s "
                 "Reason: %s", recipe.name, str(e))
 
+    def request_build_by_branch(self,
+                                branches: List[str],
+                                force: bool = False,
+                                dry_run: bool = False) -> List[object]:
+        """Request a build for the recipe(s) associated to the git branch(es).
+
+        :param branches: a list of branches to match to find the recipes.
+        :param force: if True, the build is requested even if there is a valid
+                      build.
+        :param dry_run: if True, the request for building is logged, but not
+                        submitted to Launchpad.
+        :returns: list of builds
+        """
+        builds = []
+        for recipe in self._find_recipes(branches):
+
+            build = self.request_build_by_recipe(recipe, force, dry_run)
+            if build:
+                builds.append(build)
+                print(f'{self.charmhub_name}: New build requested '
+                      f'{build.web_link}')
+
+        return builds
+
+    def request_build_by_recipe(self, recipe, force, dry_run) -> object:
+        """Request a build for the recipe.
+
+        :param recipe: recipe to request the build for.
+        :param force: if True, no checks are made to detect if the build is
+                      needed.
+        :returns: the build object or None when no build was requested.
+        """
+        build = None
+        do_build = False
+        if force:
+            logger.debug('Forcing build of recipe %s', recipe)
+            do_build = True
+        elif not self.is_build_valid(recipe.builds[0]):
+            logger.debug('Build %s is no valid, so request new build',
+                         recipe.builds[0])
+            do_build = True
+
+        if do_build:
+            if dry_run:
+                print(f'{self.charmhub_name}: New build needed (dry-run)')
+            else:
+                # the recipe build channels need to be passed when requesting
+                # the build, otherwise the build is made without the overrides
+                # that the recipe has defined.
+                build = recipe.requestBuilds(
+                    channels=recipe.auto_build_channels
+                )
+        else:
+            logger.debug('Not required to build: %s', recipe)
+
+        return build
+
+    def is_build_valid(self, build) -> bool:
+        """Determine if the build is valid.
+
+        A valid build a build that meets the following criteria:
+        - it's associated recipe has the attribute can_upload_to_store set to
+          True, but the build has no store_upload_revision set, and the build
+          is not in progress (states: currently build, uploading build or needs
+          building)
+        - the build state is in any of: 'Failed to build', 'Failed to upload'.
+        - the associated recipe is stale.
+        """
```

Comment:
Nit: the :param: and :returns: are missing here, please.

Reply:
ack

Reply:
commit [1761ec1](https://github.com/openstack-charmers/charmhub-lp-tools/pull/12/commits/1761ec1e7e298465b5b1ddead446181e881a8852) addresses this.